### PR TITLE
Respect LFS File Lock on UI

### PR DESCRIPTION
**Pull request by blueworrybear**
This PR is to update #8687.

Edit/Delete buttons are disabled if the file is currently locked by LFS and the user is not the lock owner.
The locker's name will be also shown in the file information area.

Lock/Unlock with UI is not implemented in this PR. I think it's needed to have more discussion.

LFSLock

### CODE REVIEW

File: options/locale/locale_zh-TW.ini [hunk 1]
```diff
@@ -526,6 +526,7 @@ editor.edit_file=編輯文件
 editor.preview_changes=預覽更改
 editor.cannot_edit_non_text_files=網站介面不能編輯二進位檔案
 editor.edit_this_file=編輯文件
+editor.this_file_locked = 文件已被鎖定
```
Comment (lunny) [inline options/locale/locale_zh-TW.ini]: You should go to crowdin to change this.

File: routers/repo/blame.go [hunk 2]
```diff
@@ -94,6 +94,16 @@ func RefBlame(ctx *context.Context) {
 		log.Error("GetLatestCommitStatus: %v", err)
 	}
 
+	// Check LFS Lock
+	isLFSLock := false
+	lfsLock, err := ctx.Repo.Repository.GetTreePathLock(ctx.Repo.TreePath)
+	if err != nil {
+		ctx.ServerError("GetTreePathLock", err)
```
Comment (lunny) [inline routers/repo/blame.go]: missing `return`

File: routers/repo/view.go [hunk 3]
```diff
@@ -265,6 +266,18 @@ func renderFile(ctx *context.Context, entry *git.TreeEntry, treeLink, rawLink st
 			ctx.Data["RawFileLink"] = fmt.Sprintf("%s%s.git/info/lfs/objects/%s/%s", setting.AppURL, ctx.Repo.Repository.FullName(), meta.Oid, filenameBase64)
 		}
 	}
+	lfsLock, err := ctx.Repo.Repository.GetTreePathLock(ctx.Repo.TreePath)
+	ctx.Data["LFSLock"] = lfsLock
+	if err != nil {
+		ctx.ServerError("GetTreePathLock", err)
```
Comment (lunny) [inline routers/repo/view.go:272]: As above.

File: options/locale/locale_zh-TW.ini (hunk 1 above)
Comment (lafriks) [reply options/locale/locale_zh-TW.ini]:
```suggestion
```

File: routers/repo/blame.go (hunk 2 above)
Comment (lafriks) [inline routers/repo/blame.go]:
```suggestion
		ctx.ServerError("GetTreePathLock", err)
		return
```

File: routers/repo/blame.go [hunk 4]
```diff
@@ -94,6 +94,16 @@ func RefBlame(ctx *context.Context) {
 		log.Error("GetLatestCommitStatus: %v", err)
 	}
 
+	// Check LFS Lock
+	isLFSLock := false
```
Comment (lafriks) [inline routers/repo/blame.go]: no need for this variable

File: routers/repo/blame.go [hunk 5]
```diff
@@ -119,8 +129,13 @@ func RefBlame(ctx *context.Context) {
 	ctx.Data["IsBlame"] = true
 
 	if ctx.Repo.CanEnableEditor() {
-		ctx.Data["CanDeleteFile"] = true
-		ctx.Data["DeleteFileTooltip"] = ctx.Tr("repo.editor.delete_this_file")
+		if isLFSLock {
```
Comment (lafriks) [inline routers/repo/blame.go]: just use this condition here:
```suggestion
		if lfsLock != nil && lfsLock.OwnerID != ctx.User.ID {
```

File: routers/repo/blame.go [hunk 6]
```diff
@@ -94,6 +94,16 @@ func RefBlame(ctx *context.Context) {
 		log.Error("GetLatestCommitStatus: %v", err)
 	}
 
+	// Check LFS Lock
+	isLFSLock := false
+	lfsLock, err := ctx.Repo.Repository.GetTreePathLock(ctx.Repo.TreePath)
```
Comment (lafriks) [inline routers/repo/blame.go]: These changes should be moved inside `if ctx.Repo.CanEnableEditor() {` below (line 132). No need to read and check locks if editor can not be enabled anyway

**Review by lafriks — CHANGES_REQUESTED** [review]

**Comment by blueworrybear** [comment]
@lunny @lafriks 
Thanks for the review, I am changing it and will push again later.

File: options/locale/locale_zh-TW.ini (hunk 1 above)
Comment (blueworrybear) [reply options/locale/locale_zh-TW.ini]: This line is removed.

File: routers/repo/blame.go (hunk 2 above)
Comment (blueworrybear) [reply routers/repo/blame.go]: `retuern` is added.

File: routers/repo/view.go (hunk 3 above)
Comment (blueworrybear) [reply routers/repo/view.go:272]: Added too.

File: routers/repo/blame.go (hunk 2 above)
Comment (blueworrybear) [reply routers/repo/blame.go]: `return` is added.

File: routers/repo/blame.go (hunk 4 above)
Comment (blueworrybear) [reply routers/repo/blame.go]: This variable is removed

File: routers/repo/blame.go (hunk 5 above)
Comment (blueworrybear) [reply routers/repo/blame.go]: Done.

File: routers/repo/blame.go (hunk 6 above)
Comment (blueworrybear) [reply routers/repo/blame.go]: Done.

**Review by 6543 — APPROVED** [review]

File: routers/repo/view.go [hunk 7]
```diff
@@ -265,6 +265,17 @@ func renderFile(ctx *context.Context, entry *git.TreeEntry, treeLink, rawLink st
 			ctx.Data["RawFileLink"] = fmt.Sprintf("%s%s.git/info/lfs/objects/%s/%s", setting.AppURL, ctx.Repo.Repository.FullName(), meta.Oid, filenameBase64)
 		}
 	}
+	// Check LFS Lock
+	lfsLock, err := ctx.Repo.Repository.GetTreePathLock(ctx.Repo.TreePath)
```
Comment (lafriks) [inline routers/repo/view.go:269]: We should probably also not check this if file is not LFS file

Comment (blueworrybear) [reply routers/repo/view.go:269]: @lafriks 
As @davidsvantesson mentioned in #8687, `git lfs can be used without storing the file as a git lfs file`. Therefore, a file with LFS lock may not have `LFSMetaObject` and it's `ctx.Data["IsLFSFile"]` will be false.

So it seems to me that we should always check if a file has LFSLock. If there is anything I misunderstand, please let me know. 😄

Comment (lafriks) [reply routers/repo/view.go:269]: but I could still push changes to that file from git client and LFS lock would be ignored imho, or am I wrong?

Comment (zeripath) [reply routers/repo/view.go:269]: It's dependent on the status of the lockable attribute - which can be in .gitattributes as well as .git/info/attributes (so you can't just rely on the contents of .gitattributes.)

I think it's one of those things whereby you could add a simple marker to the locked files to show that they're not marked by .gitattributes?

I'll think about it and consider adding a marker to #8726

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
Btw does it take into account selected branch?

**Comment by zeripath** [comment]
@lafriks LFS locks are pan repository not branch. They don't take account of the branch.

@lafriks @blueworrybear I can add .gitattributes checking to this in #8726 if we merge this (likely) before that's done and it's wanted.

**Pull request merged**